zerver/lib/actions.py

```diff
@@ -463,18 +463,21 @@ def process_new_human_user(user_profile: UserProfile,
                 prereg_user.referred_by,
                 _("{user} accepted your invitation to join Zulip!").format(user=f"{user_profile.full_name} <`{user_profile.email}`>")
             )
-    # Mark any other PreregistrationUsers that are STATUS_ACTIVE as
+    # Mark any other PreregistrationUsers in the realm that are STATUS_ACTIVE as
     # inactive so we can keep track of the PreregistrationUser we
-    # actually used for analytics
-    if prereg_user is not None:
+    # actually used for analytics.
+    # In the special case of realm creation, there can be no additional PreregistrationUser
+    # for us to want to modify.
+    if prereg_user is not None and not realm_creation:
         PreregistrationUser.objects.filter(
-            email__iexact=user_profile.delivery_email).exclude(id=prereg_user.id)\
+            email__iexact=user_profile.delivery_email, realm=user_profile.realm).exclude(id=prereg_user.id)\
             .update(status=confirmation_settings.STATUS_REVOKED)
 
         if prereg_user.referred_by is not None:
             notify_invites_changed(user_profile)
-    else:
-        PreregistrationUser.objects.filter(email__iexact=user_profile.delivery_email)\
+    elif prereg_user is None:
```

Comment:
Should this just be nesting inside a `if prereg_user is not None` block that has an `if realm_creation` feature in it?

May be cleaner to extract to a function `revoke_preregistration_users(delivery_email, realm: Optional[Realm])`?  And the semantics would be `realm=None` iff `realm_creation`?

Reply:
As explained above in https://github.com/zulip/zulip/pull/17275#discussion_r575282584, `if realm_creation` is not an allowable scenario without a prereg_user. Extracting the logic for revoking PreregistrationUsers seems like a great idea to clean things up, so pushed a commit.

Reply:
Cool.  I think we can do a simpler if/else sequence here, more like this:

```
def revoke_preregistration_users(                                                                    
    created_user_profile: UserProfile,                                                               
    used_preregistration_user: Optional[PreregistrationUser],                                        
    realm_creation: bool                                                                             
) -> None:                                                                                           
    if used_preregistration_user is None:              
        assert not realm_creation, "realm_creation should only happen with a PreregistrationUser"
        return
                                                                                                     
    # Mark any other PreregistrationUsers in the realm that are STATUS_ACTIVE as                     
    # inactive so we can keep track of the PreregistrationUser we                                    
    # actually used for analytics.                                                                   
    if realm_creation:                                                                               
        # In the special case of realm creation, we just need                                        
        PreregistrationUser.objects \                                                                
            .filter(email__iexact=created_user_profile.delivery_email, realm=prereg_user.realm) \                                                                                               
            .update(status=confirmation_settings.STATUS_REVOKED)                                     
    else:                                                                                            
        PreregistrationUser.objects \                                                                
            .filter(email__iexact=created_user_profile.delivery_email, realm=prereg_user.realm) \                                                                                               
            .exclude(id=used_preregistration_user.id) \                                              
            .update(status=confirmation_settings.STATUS_REVOKED)                                     
```

Reply:
Yup, pushed a commit